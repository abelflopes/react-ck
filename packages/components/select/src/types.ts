import { type InputProps } from "@react-ck/input";
import { type MenuItemProps } from "@react-ck/provisional";

export type UserValue = string | string[] | undefined;

export type SelectedValues = string[];

export type ChangeHandler = (value: string, mode: "select" | "deselect") => void;

export interface SelectProps
  extends Omit<React.HTMLAttributes<HTMLElement>, "onChange" | "value" | "defaultValue"> {
  skin?: InputProps["skin"];
  placeholder?: InputProps["placeholder"];
  children?: React.ReactNode;
  search?: {
    placeholder: string;
    emptyStateMessage: (value: string) => React.ReactNode;
  };
  onChange?: (e: React.ChangeEvent<HTMLSelectElement>, selectedValues: SelectedValues) => void;
  value?: UserValue;
  defaultValue?: UserValue;
  name?: React.SelectHTMLAttributes<HTMLSelectElement>["name"];
  multiple?: React.SelectHTMLAttributes<HTMLSelectElement>["multiple"];
  displayValueFormatter?: (data: {
    displayValue: string;
    selectedValues: string[];
    childrenData: SelectChildrenData[];
  }) => React.ReactNode;
}

export interface SelectOptionProps extends Omit<MenuItemProps, "skin"> {
  value?: string;
  disabled?: boolean;
  selected?: boolean;
}

export type SelectChildrenData = {
  isSelectOption: boolean;
  selectOptionProps: SelectOptionProps | undefined;
  computedValue: string | undefined;
  textContent: string | undefined;
  element: React.ReactNode;
};
