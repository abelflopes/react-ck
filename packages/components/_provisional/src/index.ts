export * from "./image";

export * from "./avatar";

export * from "./dropdown";

export * from "./dock";

export * from "./breadcrumbs";

export * from "./scrollable-container";

export * from "./position-engine";

export * from "./top-bar";

export * from "./snackbar";

export * from "./overlay-spinner";

export * from "./menu";

export * from "./tooltip";

export * from "./popover";

export * from "./chat";

export * from "./file-uploader";

export * from "./attachment";

export * from "./snippet";
