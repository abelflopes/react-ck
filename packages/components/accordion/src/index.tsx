import React from "react";
import { List, type ListProps } from "@react-ck/list";
import { Collapse } from "@react-ck/collapse";
import styles from "./styles/index.module.scss";

export interface AccordionItem {
  header: NonNullable<React.ReactNode>;
  children: NonNullable<React.ReactNode>;
}

export interface AccordionProps extends Omit<ListProps, "items"> {
  items: AccordionItem[];
}

/**
 * An accordion is a vertically stacked list of interactive items. Each item can be "collapsed" with only a summary visible or “expanded” to show the full content for that item.
 * @param props - {@link AccordionProps}
 * @returns a React element
 */

export const Accordion = ({
  items,
  ...otherProps
}: Readonly<AccordionProps>): React.ReactElement => (
  <List
    {...otherProps}
    items={items.map(({ header, children }) => (
      <Collapse key={String(JSON.stringify(header))} header={header}>
        <div className={styles.content}>{children}</div>
      </Collapse>
    ))}
  />
);
