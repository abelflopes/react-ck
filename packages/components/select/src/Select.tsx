import React from "react";
import styles from "./styles/index.module.scss";
import classNames from "classnames";
import { FormField, type FormFieldProps } from "@react-ck/form-field";

export interface SelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  label?: FormFieldProps["label"];
  skin?: FormFieldProps["skin"];
  description?: FormFieldProps["description"];
  validationMessage?: FormFieldProps["validationMessage"];
}

/**
 * Select is a type of input that allows users to choose one or more options from a list of choices.
 * The options are hidden by default and revealed when a user interacts with an element. It shows the currently selected option in its default collapsed state.
 * @param props - {@link SelectProps}
 * @returns a React element
 */

export const Select = ({
  skin = "default",
  label,
  description,
  validationMessage,
  className,
  ...props
}: Readonly<SelectProps>): React.ReactElement => (
  <FormField {...{ skin, label, description, validationMessage }}>
    <select {...props} className={classNames(className, styles.root)} />
  </FormField>
);
