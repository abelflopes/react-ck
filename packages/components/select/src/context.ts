import { createContext } from "react";
import { type ChangeHandler, type SelectedValues } from "./types";

export interface SelectContextProps {
  selectedValues: SelectedValues;
  handleChange: ChangeHandler;
}

export const SelectContext = createContext<SelectContextProps | undefined>(undefined);
