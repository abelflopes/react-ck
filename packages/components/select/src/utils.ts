import React from "react";
import {
  componentToText,
  getChildrenListWithoutFragments,
  getDisplayName,
} from "@react-ck/react-utils";
import {
  type SelectChildrenData,
  type SelectOptionProps,
  type SelectedValues,
  type UserValue,
} from "./types";

export const valueAsArray = (value: UserValue): SelectedValues =>
  value ? (value instanceof Array ? value : [value]) : [];

export const getChildrenData = (children: React.ReactNode): SelectChildrenData[] =>
  getChildrenListWithoutFragments(children).map<SelectChildrenData>((i) => {
    if (getDisplayName(i) === "SelectOption" && React.isValidElement<SelectOptionProps>(i)) {
      if (!("children" in i.props || "value" in i.props))
        throw new Error("SelectOption has no computable value");

      const { value } = i.props;
      const textContent = componentToText(i);
      const computedValue = value ?? textContent;

      return {
        isSelectOption: true,
        element: i,
        selectOptionProps: i.props,
        searchText: textContent ?? value,
        computedValue,
      };
    }

    return {
      isSelectOption: false,
      element: i,
      searchText: undefined,
      computedValue: undefined,
      selectOptionProps: undefined,
    };
  });
