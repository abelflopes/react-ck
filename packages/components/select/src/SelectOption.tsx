import { Menu } from "@react-ck/provisional";
import React, { useContext, useMemo } from "react";
import { SelectContext } from "./context";
import { componentToText, DISPLAY_NAME_ATTRIBUTE } from "@react-ck/react-utils";
import { type SelectOptionProps } from "./types";

const SelectOption = ({
  value,
  disabled,
  children,
  ...props
}: Readonly<SelectOptionProps>): React.ReactElement => {
  const context = useContext(SelectContext);

  const childrenText = useMemo(() => componentToText(children), [children]);

  const computedValue = useMemo(() => {
    const v = value ?? childrenText;

    if (!v || v.length === 0) throw new Error("Select option has no value");

    return v;
  }, [childrenText, value]);

  const displayChildren = useMemo(() => children ?? value, [children, value]);

  const isCurrentlySelected = context?.selectedValues.includes(computedValue);

  return (
    <Menu.Item
      {...props}
      disabled={disabled}
      skin={isCurrentlySelected ? "primary" : "default"}
      onClick={() => {
        if (disabled) return;

        context?.handleChange(computedValue, isCurrentlySelected ? "deselect" : "select");
      }}>
      {displayChildren}
    </Menu.Item>
  );
};

SelectOption[DISPLAY_NAME_ATTRIBUTE] = "SelectOption";

export { SelectOption };
