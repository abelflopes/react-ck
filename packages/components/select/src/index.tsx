import styles from "./styles/index.module.scss";
import React, { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { SelectOption } from "./SelectOption";
import { Dropdown, Menu } from "@react-ck/provisional";
import { Input } from "@react-ck/input";
import classNames from "classnames";
import { useNextRender } from "@react-ck/react-utils";
import { EmptyState } from "@react-ck/empty-state";
import { getChildrenData, valueAsArray } from "./utils";
import { type SelectProps, type ChangeHandler, type SelectOptionProps } from "./types";
import { SelectContext, type SelectContextProps } from "./context";

/**
 * Select is a type of input that allows users to choose one or more options from a list of choices.
 * The options are hidden by default and revealed when a user interacts with an element. It shows the currently selected option in its default collapsed state.
 * @param props - {@link SelectProps}
 * @returns a React element
 */

// eslint-disable-next-line max-lines-per-function -- TODO: fix
const Select = ({
  skin = "default",
  placeholder,
  children,
  className,
  onFocus,
  onBlur,
  search: searchOptions,
  onChange: selectOnChange,
  name: selectName,
  value: userValue,
  multiple: selectMultiple,
  defaultValue,
  ...props
}: Readonly<SelectProps>): React.ReactElement => {
  const onNextRender = useNextRender();
  const searchRef = useRef<HTMLInputElement>(null);
  const selectRef = useRef<HTMLSelectElement>(null);
  const rootElRef = useRef<HTMLDivElement>(null);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const blurTimeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);
  const [internalValue, setInternalValue] = useState(userValue ?? defaultValue);
  const [open, setOpen] = useState(false);
  const [search, setSearch] = useState("");

  /** Children mapped to ChildrenData object to facilitate operations */
  const childrenData = useMemo(() => getChildrenData(children), [children]);

  /** Options list children filtered by user search  */
  const filteredOptions = useMemo(
    () =>
      childrenData
        .filter(
          (i) =>
            !search.length ||
            (i.textContent && i.textContent.toLowerCase().includes(search.toLowerCase())),
        )
        .map((i) => i.element),
    [childrenData, search],
  );

  /** Returns the internal value always as an array to facilitate operations  */
  const selectedValuesList = useMemo(() => valueAsArray(internalValue ?? []), [internalValue]);

  const updateInternalValue = useCallback<ChangeHandler>(
    (value, mode) => {
      setInternalValue((v) => {
        if (!selectMultiple && mode === "select") return value;
        else if (!selectMultiple && mode === "deselect") return undefined;
        else if (mode === "deselect" && v !== undefined)
          return valueAsArray(v).filter((i) => i !== value);
        else if (v !== undefined) return [...valueAsArray(v), value];
        return [value];
      });
    },
    [selectMultiple],
  );

  const handleFocus = useCallback(() => {
    if (blurTimeoutRef.current) clearTimeout(blurTimeoutRef.current);
    setOpen(true);
  }, []);

  const handleBlur = useCallback(() => {
    if (blurTimeoutRef.current) clearTimeout(blurTimeoutRef.current);
    blurTimeoutRef.current = setTimeout(() => {
      setOpen(false);
    }, 150);
  }, []);

  /**
   * Handle change events trigger by the user
   * Also replicates the event on the native select
   */
  const handleChange = useCallback<ChangeHandler>(
    (value, mode) => {
      updateInternalValue(value, mode);

      if (!selectMultiple) setOpen(false);

      onNextRender(() => {
        if (!selectRef.current)
          throw new Error("Select: unable to dispatch event to native element");

        selectRef.current.dispatchEvent(
          new Event("change", {
            bubbles: true,
          }),
        );
      });
    },
    [onNextRender, selectMultiple, updateInternalValue],
  );

  /** Context object propagated to select options  */
  const contextValue = useMemo<SelectContextProps>(
    () => ({
      handleChange,
      selectedValues: selectedValuesList,
    }),
    [handleChange, selectedValuesList],
  );

  /** Actions to do when dropdown opens  */
  useEffect(() => {
    if (!open) return;

    // Focus on search input
    if (search) {
      onNextRender(() => {
        searchRef.current?.focus();
      });
    }
  }, [onNextRender, open, search]);

  /** Actions to do when dropdown closes  */
  useEffect(() => {
    if (!open) return;
    // Clear search
    setSearch("");
  }, [open]);

  /** Set internal state with user provided value  */
  useEffect(() => {
    setInternalValue(userValue);
  }, [userValue]);

  /**Update internal values when children selected attribute changes */
  useEffect(() => {
    childrenData.forEach(({ selectOptionProps, computedValue, isSelectOption }) => {
      const { selected } = selectOptionProps || {};
      // skip if item is not a select option or the attribute is not set
      if (!isSelectOption || selected === undefined || computedValue === undefined) return;

      updateInternalValue(computedValue, selected ? "select" : "deselect");
    });
  }, [childrenData, updateInternalValue]);

  return (
    <>
      <select
        ref={selectRef}
        name={selectName}
        multiple={selectMultiple}
        defaultValue={defaultValue}
        value={internalValue}
        hidden
        onChange={(e) => {
          selectOnChange?.(e, selectedValuesList);
        }}>
        {selectedValuesList.map((i) => (
          <option key={i} value={i}>
            {i}
          </option>
        ))}
      </select>

      <div
        {...props}
        ref={rootElRef}
        role="button"
        tabIndex={0}
        className={classNames(styles.root, styles[`skin_${skin}`], className)}
        onFocus={(e) => {
          handleFocus();
          onFocus?.(e);
        }}
        onBlur={(e) => {
          handleBlur();
          onBlur?.(e);
        }}>
        {selectedValuesList.length > 0 &&
          childrenData
            .filter((i) => i.computedValue && selectedValuesList.includes(i.computedValue))
            .map((i) => i.textContent)
            .join(", ")}
        {selectedValuesList.length === 0 && (
          <span className={styles.placeholder}>{placeholder}</span>
        )}
      </div>

      <Dropdown
        anchorRef={rootElRef}
        open={open}
        spacing="s"
        rootRef={dropdownRef}
        excludeAutoPosition={["left", "right", "start", "end", "full"]}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onClose={() => {
          setOpen(false);
        }}>
        <Menu className={styles.menu}>
          {searchOptions ? (
            <>
              <Input
                rootRef={searchRef}
                value={search}
                type="search"
                placeholder={searchOptions.placeholder}
                skin="ghost"
                className={styles.search_input}
                onChange={(e) => {
                  setSearch(e.target.value);
                }}
              />
              <Menu.Divider />
            </>
          ) : null}

          <SelectContext.Provider value={contextValue}>{filteredOptions}</SelectContext.Provider>

          {searchOptions && filteredOptions.length === 0 ? (
            <EmptyState>
              <span>{searchOptions.emptyStateMessage(search)}</span>
              {/* TODO: fix empty state flex */}
            </EmptyState>
          ) : null}
        </Menu>
      </Dropdown>
    </>
  );
};

Select.Option = SelectOption;

export { Select, type SelectProps, type SelectOptionProps };
