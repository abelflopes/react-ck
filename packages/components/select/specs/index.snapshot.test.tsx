import React from "react";
import { Select } from "../src/index";
import renderer from "react-test-renderer";

describe("Snapshot Select", () => {
  test("renders correctly", async () => {
    const tree = renderer.create(<Select />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
