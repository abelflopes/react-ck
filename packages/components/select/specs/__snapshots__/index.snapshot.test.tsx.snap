// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Select renders correctly 1`] = `
[
  <select
    hidden={true}
    onChange={[Function]}
  />,
  <div
    className="root skin_default"
    onFocus={[Function]}
    role="button"
    tabIndex={0}
  />,
]
`;
